### ISSUE TITLE
Failed to bind properties under 'server.tomcat.basedir' to java.io.File for 1.5.11

### ISSUE THREAD

**Issue by hdconradi** [spring-projects/spring-boot#12786]
When updating from 1.5.10 -> 1.5.11 and using `server.tomcat.basedir=.` or `server.tomcat.basedir=\.`
the application failed to start with the error

```
Binding to target org.springframework.boot.autoconfigure.web.ServerProperties@66948673 failed:

    Property: server.tomcat.basedir
    Value: ./
    Reason: Failed to convert property value of type 'java.lang.String' to required type 'java.io.File' for property 'tomcat.basedir'; nested exception is java.lang.IllegalArgumentException: Could not retrieve file for class path resource []: class path resource [] cannot be resolved to absolute file path because it does not reside in the file system: jar:file:/opt/icw/product/c2id-facade-1.4-SNAPSHOT/c2id-facade.jar!/BOOT-INF/classes!/
```
I have reproduced the failure with the `spring-boot-sample-web-secure` project, by adding the property:
```
diff --git a/spring-boot-samples/spring-boot-sample-web-secure/src/main/resources/application.properties b/spring-boot-samples/spring-boot-sample-web-secure/src/main/resources/application.properties
index e1935fa698..2c0ba1c4dd 100644
--- a/spring-boot-samples/spring-boot-sample-web-secure/src/main/resources/application.properties
+++ b/spring-boot-samples/spring-boot-sample-web-secure/src/main/resources/application.properties
@@ -1,3 +1,5 @@
+server.tomcat.basedir=./
+
 spring.thymeleaf.cache: false
 security.basic.enabled: false
 # demo only:
```
It is important to notice that the project continues to build fine.
However, when executing the jar with e.g.:
```
java -jar target/spring-boot-sample-web-secure-1.5.11.RELEASE.jar
```
the above error occurs.
This is probably due to the fact that in the `ClassPathResource#getURL` the url has the value
`jar:file:/home/<some-path>/target/spring-boot-sample-web-secure-1.5.11.RELEASE.jar!/BOOT-INF/classes!/` while the 'jar'-prefix is missing starting the application in the tests.

This might be related to #12163. 
That is a spring-boot 2 bug, but the description reads very similar.
It is in particular true that a value `server.tomcat.basedir=../spring-boot-sample-web-secure` works.

**Comment by wilkinsona** [comment]
#12163 only affected 2.0.0 so I don't think it's relevant here. I also unable to reproduce the problem using 1.5.11 with either value that you have described above.

Can you please provide a minimal sample that reproduces the problem?

**Comment by hdconradi** [comment]
I have updated the description accordingly.

**Comment by hdconradi** [comment]
In 1.5.12 the behavior has not changed (I know it wasn't supposed to, but would like to take the opportunity to point out an even simpler way to reproduce it):
```
java -Dserver.tomcat.basedir=. -jar target/spring-boot-sample-web-secure-1.5.12.RELEASE.jar
```
still leads to the behavior described above.

**Comment by wilkinsona** [comment]
Thank you. I think the thing that was missing from your original description (now lost since your edit) was that the problem only occurs when packaged as a jar. I believe it can be avoided by using `file:.` instead of just `.`.

**Comment by hdconradi** [comment]
You are correct: `-Dserver.tomcat.basedir=file:.` works.

**Comment by wilkinsona** [comment]
1.5.10 uses the `ConversionService` and its `ObjectToObjectConverter` to successfully turn the `.` `String` into a `File` using its `File(String)` constructor. In 1.5.11, there's a `FileEditor` available. Its availability stops the `ConversionService` from being used. Instead, an attempt is made to use the `FileEditor` to perform the conversion. It fails with the following exception:

```
java.lang.IllegalArgumentException: Could not retrieve file for class path resource []: class path resource [] cannot be resolved to absolute file path because it does not reside in the file system: jar:file:/Users/awilkinson/dev/workspaces/spring/spring-boot/1.5.x-old/gh-12786/target/gh-12786-0.0.1-SNAPSHOT.jar!/BOOT-INF/classes!/
	at org.springframework.beans.propertyeditors.FileEditor.setAsText(FileEditor.java:107)
	at org.springframework.beans.TypeConverterDelegate.doConvertTextValue(TypeConverterDelegate.java:470)
	at org.springframework.beans.TypeConverterDelegate.doConvertValue(TypeConverterDelegate.java:443)
	at org.springframework.beans.TypeConverterDelegate.convertIfNecessary(TypeConverterDelegate.java:200)
	at org.springframework.beans.AbstractNestablePropertyAccessor.convertIfNecessary(AbstractNestablePropertyAccessor.java:588)
	at org.springframework.beans.AbstractNestablePropertyAccessor.convertForProperty(AbstractNestablePropertyAccessor.java:615)
	at org.springframework.beans.AbstractNestablePropertyAccessor.processLocalProperty(AbstractNestablePropertyAccessor.java:462)
	at org.springframework.beans.AbstractNestablePropertyAccessor.setPropertyValue(AbstractNestablePropertyAccessor.java:290)
	at org.springframework.beans.AbstractNestablePropertyAccessor.setPropertyValue(AbstractNestablePropertyAccessor.java:278)
	at org.springframework.boot.bind.RelaxedDataBinder$RelaxedBeanWrapper.setPropertyValue(RelaxedDataBinder.java:699)
	at org.springframework.beans.AbstractPropertyAccessor.setPropertyValues(AbstractPropertyAccessor.java:95)
	at org.springframework.validation.DataBinder.applyPropertyValues(DataBinder.java:860)
	at org.springframework.validation.DataBinder.doBind(DataBinder.java:756)
	at org.springframework.boot.bind.RelaxedDataBinder.doBind(RelaxedDataBinder.java:128)
	at org.springframework.validation.DataBinder.bind(DataBinder.java:741)
	at org.springframework.boot.bind.PropertiesConfigurationFactory.doBindPropertiesToTarget(PropertiesConfigurationFactory.java:287)
	at org.springframework.boot.bind.PropertiesConfigurationFactory.bindPropertiesToTarget(PropertiesConfigurationFactory.java:250)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:331)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:292)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:409)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1622)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:308)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:519)
	at org.springframework.boot.context.embedded.EmbeddedServletContainerCustomizerBeanPostProcessor.getCustomizers(EmbeddedServletContainerCustomizerBeanPostProcessor.java:82)
	at org.springframework.boot.context.embedded.EmbeddedServletContainerCustomizerBeanPostProcessor.postProcessBeforeInitialization(EmbeddedServletContainerCustomizerBeanPostProcessor.java:72)
	at org.springframework.boot.context.embedded.EmbeddedServletContainerCustomizerBeanPostProcessor.postProcessBeforeInitialization(EmbeddedServletContainerCustomizerBeanPostProcessor.java:59)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:409)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1622)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:308)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.getEmbeddedServletContainerFactory(EmbeddedWebApplicationContext.java:199)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.createEmbeddedServletContainer(EmbeddedWebApplicationContext.java:162)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.onRefresh(EmbeddedWebApplicationContext.java:134)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:537)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1118)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1107)
	at com.example.demo.Gh12786Application.main(Gh12786Application.java:10)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
```

The `FileEditor` is available due to the changes made in https://github.com/spring-projects/spring-boot/commit/3db5c70b582e3c899aea2fd5e0f5fead4d60b689 to fix https://github.com/spring-projects/spring-boot/issues/11569.

**Comment by hdconradi** [comment]
I expected something like that ;-)
Is it planned to restore the backward compatibility of this feature in a future 5.x release?

**Comment by wilkinsona** [comment]
@snicoll I’m not sure how to fix this without replacing this regression with a regression of your fix for #11569. Any suggestions?

**Comment by hdconradi** [comment]
Thanks a lot.
